File: C/hardware/qube_servo2_usb_control_example/qube_servo2_usb_control_example.c
//////////////////////////////////////////////////////////////////
//
// qube_servo2_usb_control_example.c - C file
//
// This example demonstrates how to do proportional control using the Quanser HIL SDK.
// It is designed to control Quanser's Qube Servo2 USB experiment. The position of the
// motor is read from encoder channel 0. Analog output channel 0 is used to
// drive the motor.
//
// This example runs until Ctrl+C is pressed.
//
// This example demonstrates the use of the following functions:
//    hil_open
//    hil_set_encoder_counts
//    hil_task_create_encoder_reader
//    hil_task_start
//    hil_task_read_encoder
//    hil_write_analog
//    hil_task_stop
//    hil_task_delete
//    hil_close
//
// Copyright (C) 2018 Quanser Inc.
//////////////////////////////////////////////////////////////////
    
#include "qube_servo2_usb_control_example.h"

static int stop = 0;

void signal_handler(int signal)
{
    stop = 1;
}

int main(int argc, char* argv[])
{
    static const char board_type[]       = "qube_servo2_usb";
    static const char board_identifier[] = "0";
    static char       message[512];

    qsigaction_t action;
    t_card board;
    t_int  result;

    /* Catch Ctrl+C so application may shut down cleanly */
    action.sa_handler = signal_handler;
    action.sa_flags   = 0;
    qsigemptyset(&action.sa_mask);

    qsigaction(SIGINT, &action, NULL);

    result = hil_open(board_type, board_identifier, &board);
    if (result == 0)
    {
        const t_uint32 samples              = -1; /* read continuously */
        const t_uint32 analog_channel       = 0;
        const t_uint32 digital_channel      = 0;
        const t_uint32 encoder_channel      = 0;
        const t_double frequency            = 1000;
        const t_double sine_frequency       = 0.5; /* frequency of command signal */
        const t_uint32 samples_in_buffer    = (t_uint32)(0.1 * frequency);
        const t_double period               = 1.0 / frequency;

        t_int32   count;
        t_double  voltage;
        t_boolean enable;
        t_int     samples_read = 0;
        t_task    task;

        printf("This example controls the Quanser Qube Servo2 USB experiment at %g Hz.\n", frequency);
        
        printf("Press CTRL-C to stop the controller.\n\n");
        
        do
        {
            qsched_param_t scheduling_parameters;

            /* Make sure the motor voltage is zero */
            voltage = 0;
            result = hil_write_analog(board, &analog_channel, 1, &voltage);
            if (result < 0)
                break;

            /* Reset the encoder counter to zero */
            count = 0;
            result = hil_set_encoder_counts(board, &encoder_channel, 1, &count);
            if (result < 0)
                break;

            /* Enable the motor */
            enable = 1;
            result = hil_write_digital(board, &digital_channel, 1, &enable);
            if (result < 0)
                break;

            /* Increase our thread priority so we get better sample time performance */
            scheduling_parameters.sched_priority = qsched_get_priority_max(QSCHED_FIFO);
            qthread_setschedparam(qthread_self(), QSCHED_FIFO, &scheduling_parameters);

            /* Create a task to read the encoder. The task will be used to time the control loop */
            result = hil_task_create_encoder_reader(board, samples_in_buffer, &encoder_channel, 1, &task);
            if (result < 0)
                break;

            /* Start the task, which will cause it to start reading the encoder at the given rate */
            result = hil_task_start(task, HARDWARE_CLOCK_0, frequency, samples);
            if (result == 0)
            {
                const t_double gain = 0.3;

                t_double position;
                t_double error;
                t_double command = 0;
                t_double time = 0;

                /* The proportional control loop */
                samples_read = hil_task_read_encoder(task, 1, &count);
                while (samples_read > 0 && stop == 0)
                {
                    position = count * 360 / 2048;     /* convert counts to degrees */
                    error    = command - position;     /* compute error in position */
                    voltage  = gain * error;           /* apply proportional control */
                        
                    hil_write_analog(board, &analog_channel, 1, &voltage);
                    
                    /* Compute command signal for next sampling instant */
                    time += period;
                    command = 45 * sin(2*M_PI*sine_frequency*time);
                    samples_read = hil_task_read_encoder(task, 1, &count);
                }

                hil_task_stop(task);
            }
            else
            {
                msg_get_error_message(NULL, result, message, ARRAY_LENGTH(message));
                printf("Unable to start task. %s Error %d.\n", message, -result);
            }

            /* Turn off motor */
            voltage = 0;
            hil_write_analog(board, &analog_channel, 1, &voltage);

            enable = 0;
            hil_write_digital(board, &digital_channel, 1, &enable);

            if (samples_read < 0)
            {
                msg_get_error_message(NULL, samples_read, message, ARRAY_LENGTH(message));
                printf("Unable to read encoder channel. %s Error %d.\n", message, -samples_read);
            }
            else
            {
                printf("\nController has been stopped. Press Enter to continue.\n");
                getchar(); /* absorb Ctrl-C */
            }

            hil_task_delete(task);
        } while (false);

        if (result < 0)
        {
            msg_get_error_message(NULL, result, message, ARRAY_LENGTH(message));
            printf("%s Error %d.\n", message, -result);
        }

        hil_close(board);
    }
    else
    {
        msg_get_error_message(NULL, result, message, ARRAY_LENGTH(message));
        printf("Unable to open board. %s Error %d.\n", message, -result);
    }

    return 0;
}

File: python/hardware/qube_servo2_usb_control_example/qube_servo2_usb_control_example.py
#################################################################
#
# qube_servo2_usb_control_example.py - Python file
#
# This example demonstrates how to do proportional control using the Quanser HIL SDK.
# It is designed to control Quanser's Qube Servo2 USB experiment. The position of the
# motor is read from encoder channel 0. Analog output channel 0 is used to
# drive the motor.
#
# This example runs until Ctrl+C is pressed.
#
# This example demonstrates the use of the following functions:
#    HIL.open
#    HIL.set_encoder_counts
#    HIL.task_create_encoder_reader
#    HIL.task_start
#    HIL.task_read_encoder
#    HIL.write_analog
#    HIL.task_stop
#    HIL.task_delete
#    HIL.close
#
# Copyright (C) 2023 Quanser Inc.
#################################################################
    
from quanser.hardware import HIL, HILError, Clock
import math
import signal
import sys
import array as arr

stop = False

def signal_handler(signum, frame):
    global stop
    stop = True

# Register a Ctrl+C handler
signal.signal(signal.SIGINT, signal_handler)

board_type = "qube_servo2_usb"
board_identifier = "0"

try:
    card = HIL()
    card.open(board_type, board_identifier)

    try:
        samples           = HIL.INFINITE # write continuously
        samples_to_write  = 1
        frequency         = 1000
        sine_frequency    = 0.5 # frequency of command signal
        samples_in_buffer = math.ceil(0.1 * frequency)
        period            = 1 / frequency

        analog_channels   = arr.array("i", [0])
        digital_channels  = arr.array("i", [0])
        encoder_channels  = arr.array("i", [0])

        num_analog_channels  = len(analog_channels)
        num_encoder_channels = len(encoder_channels)
        num_digital_channels = len(digital_channels)

        voltages = arr.array("d", [0.0] * num_analog_channels)
        counts   = arr.array("i", [0] * num_encoder_channels)
        states   = arr.array("b", [1] * num_digital_channels)

        print("This example controls the Quanser Qube Servo2 USB experiment at %g Hz." % frequency)
        
        print("Press CTRL-C to stop the controller.\n")

        # Make sure the motor voltage is zero
        card.write_analog(analog_channels, num_analog_channels, voltages)

        # Reset the encoder counts to zero
        card.set_encoder_counts(encoder_channels, num_encoder_channels, counts)

        # Enable the motor
        card.write_digital(digital_channels, num_digital_channels, states)

        # Create a task to read the encoder. The task will be used to time the control loop
        task = card.task_create_encoder_reader(samples_in_buffer, encoder_channels, num_encoder_channels)
          
        # Start the task, which will cause it to start reading the encoder at the given rate */
        card.task_start(task, Clock.HARDWARE_CLOCK_0, frequency, samples)

        # Begin control
        gain = 0.3
        command = 0
        time = 0

        samples_read = card.task_read_encoder(task, 1, counts)
        while samples_read > 0 and stop == False:

            position    = counts[0] * 360 / 2048 # convert counts to degrees
            error       = command - position     # compute error in position
            voltages[0] = gain * error           # apply proportional control
                        
            card.write_analog(analog_channels, num_analog_channels, voltages)
                    
            # Compute command signal for next sampling instant
            time += period
            command = 45 * math.sin(2 * math.pi * sine_frequency * time)
            samples_read = card.task_read_encoder(task, 1, counts)

        card.task_stop(task)
        card.task_delete(task)

        # Turn off motor
        voltages[0] = 0.0
        card.write_analog(analog_channels, num_analog_channels, voltages)

        # Disable motor
        states[0] = 0
        card.write_digital(digital_channels, num_digital_channels, states)

    except HILError as ex:
        print("Unable to run task. %s" % ex.get_error_message())

    finally:
        card.close()

except HILError as ex:
    print("Unable to open board. %s" % ex.get_error_message())

input("Press Enter to continue.")
exit(0)
